crates/client/sync/src/fetch/fetchers.rs

```diff
@@ -277,52 +269,58 @@ async fn fetch_class_updates(
     state_update: &starknet_providers::sequencer::models::StateUpdate,
     block_id: FetchBlockId,
     provider: &SequencerGatewayProvider,
-) -> anyhow::Result<Vec<DbClassUpdate>> {
-    let missing_classes: Vec<_> = std::iter::empty()
-        .chain(
-            state_update
-                .state_diff
-                .deployed_contracts
-                .iter()
-                .map(|deployed_contract| (deployed_contract.class_hash, &Felt::ZERO)),
-        )
-        .chain(state_update.state_diff.old_declared_contracts.iter().map(|&felt| (felt, &Felt::ZERO)))
-        .chain(
-            state_update
-                .state_diff
-                .declared_classes
-                .iter()
-                .map(|declared_class| (declared_class.class_hash, &declared_class.compiled_class_hash)),
-        )
-        .collect::<HashMap<_, _>>() // unique() by key
-        .into_iter()
-        .filter_map(|(class_hash, compiled_class_hash)| {
-            match backend.contains_class(&BlockId::Tag(BlockTag::Latest), &class_hash) {
-                Ok(false) => Some(Ok((class_hash, compiled_class_hash))),
-                Ok(true) => None,
-                Err(e) => Some(Err(e)),
-            }
-        })
-        .collect::<Result<_, _>>()?;
-
-    let classes = futures::future::try_join_all(missing_classes.into_iter().map(
-        |(class_hash, &compiled_class_hash)| async move {
-            // TODO(correctness): Skip what appears to be a broken Sierra class definition (quick fix)
-            if class_hash
-                != Felt::from_hex_unchecked("0x024f092a79bdff4efa1ec86e28fa7aa7d60c89b30924ec4dab21dbfd4db73698")
-            {
-                // Fetch the class definition in parallel, retrying up to 15 times for each class
-                let (class_hash, contract_class) =
-                    retry(|| fetch_class(class_hash, block_id, provider), 15, Duration::from_secs(1)).await?;
-                Ok::<_, L2SyncError>(Some(DbClassUpdate { class_hash, contract_class, compiled_class_hash }))
-            } else {
-                Ok(None)
-            }
-        },
-    ))
-    .await?;
+) -> anyhow::Result<Vec<ClassUpdate>> {
+    let chain_id: Felt = backend.chain_config().chain_id.to_felt();
+
+    // for blocks before 2597 on mainnet new classes are not declared in the state update
+    // https://github.com/madara-alliance/madara/issues/233
+    let legacy_classes: Vec<_> = if chain_id == MAIN_CHAIN_ID && block_id.block_n() < Some(2597) {
+        let block_number = block_id.block_n().unwrap(); // Safe to unwrap because of the condition above
```

Comment:
Wouldn't it still be possible for this to be `None`? I'm assuming `None < Some(2597) == true`. If that is the case we might want to consider unwrapping to the maximum integer value of a block number.

Reply:
None for pending